App crashes when trying modify account if default transaction account no longer exist.

#### Steps to reproduce the behaviour
1. Use double entry mode.
2. Create account A and B.
3. Set B's default transaction account to A.
4. Remove account A.
5. Modify account B, crash occurred.

#### Expected behaviour
Open account modifying interface.

#### Actual behaviour
App crashed.

#### Software specifications
* GnuCash Android version: 2.1.4
* System Android version: Nougat 7.1.1
* Device type: Google Pixel

Confirmed. Stack trace:
```
java.lang.RuntimeException: Unable to start activity ComponentInfo{org.gnucash.android.devel/org.gnucash.android.ui.common.FormActivity}: java.lang.IllegalArgumentException: accounts with GUID 08b2c273aa3e96307fc6f4bbcc48563b does not exist in the db
    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2416)
    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
    at android.app.ActivityThread.access$900(ActivityThread.java:150)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
    at android.os.Handler.dispatchMessage(Handler.java:102)
    at android.os.Looper.loop(Looper.java:148)
    at android.app.ActivityThread.main(ActivityThread.java:5417)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
 Caused by: java.lang.IllegalArgumentException: accounts with GUID 08b2c273aa3e96307fc6f4bbcc48563b does not exist in the db
    at org.gnucash.android.db.adapter.DatabaseAdapter.getID(DatabaseAdapter.java:542)
    at org.gnucash.android.ui.account.AccountFormFragment.initializeViewsWithAccount(AccountFormFragment.java:395)
    at org.gnucash.android.ui.account.AccountFormFragment.onActivityCreated(AccountFormFragment.java:351)
    at android.support.v4.app.Fragment.performActivityCreated(Fragment.java:2089)
    at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1133)
    at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1290)
    at android.support.v4.app.BackStackRecord.run(BackStackRecord.java:801)
    at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1677)
    at android.support.v4.app.FragmentController.execPendingActions(FragmentController.java:388)
    at android.support.v4.app.FragmentActivity.onStart(FragmentActivity.java:604)
    at android.support.v7.app.AppCompatActivity.onStart(AppCompatActivity.java:178)
    at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1238)
    at android.app.Activity.performStart(Activity.java:6302)
    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2379)
    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
    at android.app.ActivityThread.access$900(ActivityThread.java:150) 
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
    at android.os.Handler.dispatchMessage(Handler.java:102) 
    at android.os.Looper.loop(Looper.java:148) 
    at android.app.ActivityThread.main(ActivityThread.java:5417) 
    at java.lang.reflect.Method.invoke(Native Method) 
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
```

--- a/app/src/main/java/org/gnucash/android/db/DatabaseSchema.java
+++ b/app/src/main/java/org/gnucash/android/db/DatabaseSchema.java
@@ -39,7 +39,7 @@ public class DatabaseSchema {
      * Version number of database containing accounts and transactions info.
      * With any change to the database schema, this number must increase
      */
-    public static final int DATABASE_VERSION = 13;
+    public static final int DATABASE_VERSION = 14;
 
     /**
      * Name of the database

--- a/app/src/main/java/org/gnucash/android/db/MigrationHelper.java
+++ b/app/src/main/java/org/gnucash/android/db/MigrationHelper.java
@@ -1474,4 +1474,38 @@ public class MigrationHelper {
 
         return oldVersion;
     }
+
+    /**
+     * Upgrades the database to version 14.
+     * <p>This migration makes the following changes to the database:
+     * <ul>
+     *     <li>Fixes accounts referencing a default transfer account that no longer
+     *         exists (see #654)</li>
+     * </ul>
+     * </p>
+     * @param db SQLite database to be upgraded
+     * @return New database version, 14 if migration succeeds, 13 otherwise
+     */
+    static int upgradeDbToVersion14(SQLiteDatabase db) {
+        Log.i(DatabaseHelper.LOG_TAG, "Upgrading database to version 14");
+        int oldVersion = 13;
+
+        db.beginTransaction();
+        try {
+            ContentValues contentValues = new ContentValues();
+            contentValues.putNull(AccountEntry.COLUMN_DEFAULT_TRANSFER_ACCOUNT_UID);
+            db.update(
+                    AccountEntry.TABLE_NAME,
+                    contentValues,
+                    AccountEntry.TABLE_NAME + "." + AccountEntry.COLUMN_DEFAULT_TRANSFER_ACCOUNT_UID
+                            + " NOT IN (SELECT " + AccountEntry.COLUMN_UID
+                            + "             FROM " + AccountEntry.TABLE_NAME + ")",
+                    null);
+            db.setTransactionSuccessful();
+            oldVersion = 14;
+        } finally {
+            db.endTransaction();
+        }
+        return oldVersion;
+    }
 }

--- a/app/src/main/java/org/gnucash/android/ui/account/AccountsListFragment.java
+++ b/app/src/main/java/org/gnucash/android/ui/account/AccountsListFragment.java
@@ -247,7 +247,9 @@ public class AccountsListFragment extends Fragment implements
         if (acc.getTransactionCount() > 0 || mAccountsDbAdapter.getSubAccountCount(acc.getUID()) > 0) {
             showConfirmationDialog(rowId);
         } else {
-            mAccountsDbAdapter.deleteRecord(rowId);
+            // Avoid calling AccountsDbAdapter.deleteRecord(long). See #654
+            String uid = mAccountsDbAdapter.getUID(rowId);
+            mAccountsDbAdapter.deleteRecord(uid);
             refresh();
         }
     }

